pkg/util/prefetch/reader.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prefetch
+
+import (
+	"bytes"
+	"io"
+)
+
+// Reader is a reader that prefetches data from the underlying reader.
+type Reader struct {
+	r            io.ReadCloser
+	curBufReader *bytes.Reader
+	buf          [2][]byte
+	bufIdx       int
+	bufCh        chan []byte
+	err          error // after bufCh is closed
+}
+
+// NewReader creates a new Reader.
+func NewReader(r io.ReadCloser, prefetchSize int) io.ReadCloser {
+	ret := &Reader{
+		r:     r,
+		bufCh: make(chan []byte),
+		err:   nil,
+	}
+	ret.buf[0] = make([]byte, prefetchSize/2)
+	ret.buf[1] = make([]byte, prefetchSize/2)
+	go ret.run()
+	return ret
+}
+
+func (r *Reader) run() {
+	for {
+		r.bufIdx = (r.bufIdx + 1) % 2
+		buf := r.buf[r.bufIdx]
+		n, err := r.r.Read(buf)
+		buf = buf[:n]
+		r.bufCh <- buf
```

Comment:
might block forever if we call close before read

and we should wait this routine exist on close

Reply:
https://github.com/pingcap/tidb/pull/48587/commits/b407fbf785e7b35fbfce042c8251176f34f42dc1